File: lecture_009(Recursion)/l_015_allindices.cpp
#include<iostream>
#include<vector>
using namespace std;

void input(vector<int> &arr) {
    for(int i=0; i<arr.size(); i++) {
        cin>>arr[i];
    }
}
void display(vector<int> &arr) {
    for(int i=0; i<arr.size(); i++) {
        cout<<arr[i]<<endl;
    }
}
vector<int> allIndices(vector<int> &arr, int idx, int x, int fsf) {
    if(idx==arr.size()) {
        return vector<int>(fsf,0);
    }
    if(arr[idx]==x) {
        vector<int>iarr = allIndices(arr,idx+1,x,fsf+1);
        iarr[fsf] = idx;
        return iarr;
    }
    else {
        vector<int>iarr = allIndices(arr,idx+1,x,fsf);
        return iarr;
    }
}
int main() {
    int n;
    cin>>n;
    vector<int>arr(n,0);
    input(arr);
    int d;
    cin>>d;
    vector<int>ai = allIndices(arr,0,d,0);
    display(ai);

    return 0;
}

File: lecture_009(Recursion)/l_015_allindices.java
import java.util.Scanner;
public class l_015_allindices {
    public static Scanner scn = new Scanner(System.in);
    public static void main(String[] args) {
        int n = scn.nextInt();
        int[] arr = new int[n];
        input(arr);
        int d = scn.nextInt();
        int[] ai = allIndices(arr,0,d,0);
        display(ai);
    }
    public static void display(int[] arr) {
        for(int i=0; i<arr.length; i++) {
            System.out.println(arr[i]);
        }
    }
    public static void input(int[] arr) {
        for(int i=0; i<arr.length; i++) {
            arr[i] = scn.nextInt();
        }
    }
    public static int[] allIndices(int[] arr, int idx, int x, int fsf) {
        if(idx==arr.length) {              //fsf = found so far;
            return new int[fsf];
        }
        if(arr[idx]==x) {
            int[] iarr = allIndices(arr,idx+1,x,fsf+1); 
            iarr[fsf] = idx;
            return iarr;
        }
        else {
            int[] iarr = allIndices(arr,idx+1,x,fsf);
            return iarr;
        }
    }
}
